### PR TITLE
Artemis drive 

### PR DESCRIPTION
**Pull request by soapcyst**
[Feature] [Events] [Vehicles]
## About the PR 
This Adds a drive system to the Artemis. This adds three machines

#### the interface
Manages player input and interaction with the two other machines
#### the tank
An atmos machine that intakes fuel
Filters out non-fuel gas and ejects it
#### The drive
This part of the drive is what impacts the ships performance, influences speed and fuel use.
Allows players to create parts from materials with different effects
exciter:impacts how much fuel is converted. impacts fuel use and maximum speed
casing:currently only protects the coils, more density means it will have more hp and more chem resistance means it degrades slower
coil: How the fuel is contained in the system. stronger strength means higher fuel efficiency and later on impacts malfunctions.
## Why's this needed? 
The artemis needs a fuel use and propulsion system and this is my proposal

## Changelog 

```changelog
(u)Soapcyst
(+)added a drive to the artemis
```

### CODE REVIEW

File: maps/artemis/planets.dmm [hunk 1]
```diff
@@ -766,6 +789,12 @@
 	},
 /turf/simulated/floor/black,
 /area/ship/artemis)
+"uf" = (
+/obj/machinery/atmospherics/pipe/vent{
+	dir = 4
+	},
+/turf/simulated/floor/airless/engine/vacuum,
+/area/ship/space_canvas)
```
Comment (Azrun) [inline maps/artemis/planets.dmm:797]: Why there a vent?!?

**Review by Azrun — CHANGES_REQUESTED** [review]

**Comment by deathrobotpunch** [comment]
I can’t wait for the artemis map to go public

File: maps/artemis/planets.dmm (hunk 1 above)
Comment (soapcyst) [reply maps/artemis/planets.dmm:797]: the vent is there to get rid of filtered out gasses

File: code/WorkInProgress/artemis/artemis_engine.dm [hunk 2]
```diff
@@ -0,0 +1,187 @@
+#if ENABLE_ARTEMIS
+
+/////////////////////TANK/////////////////////
+/obj/machinery/atmospherics/binary/ion_drive/plas_tank
```
[the remark this answers is not in the document]
Comment (soapcyst) [reply code/WorkInProgress/artemis/artemis_engine.dm]: changed to plasma_tank

File: _std/setup.dm [hunk 3]
```diff
@@ -75,7 +75,7 @@
 // other toggles
 
 #define FOOTBALL_MODE 1
-#define ENABLE_ARTEMIS 0
+#define ENABLE_ARTEMIS 1
```
[the remark this answers is not in the document]
Comment (soapcyst) [reply _std/setup.dm]: changed all these back to what they should be and made git ignore them locally

File: maps/artemis/planets.dmm (hunk 1 above)
Comment (Azrun) [reply maps/artemis/planets.dmm:797]: Is there a benefit from simulating that exchange instead of simply removing it?

Comment (soapcyst) [reply maps/artemis/planets.dmm:797]: No, but I not familiar with any other atmos system that just deletes it.

Comment (soapcyst) [reply maps/artemis/planets.dmm]: artemis engine has been changed to be unary and now deletes removed gas

**Comment by soapcyst** [comment]
While its still a WIP i think its in a state thats ready to be worked on by others. 

what's left to do at this point for the engine

- Create proper sprites for the engine
- finalize the component system
- implement thrusters
- tweak values so that the engine is fun and good to use.

File: code/WorkInProgress/artemis/artemis_movement_controller.dm [hunk 4]
```diff
@@ -165,7 +173,8 @@
 
 					SPAWN(ship.animation_speed)
 						ship.rotating = 0
-
+		else if(keys)
+			user.show_message("<span class='alert'>The engine isn't on!</span>")
```
Comment (Azrun) [inline code/WorkInProgress/artemis/artemis_movement_controller.dm:177]: Consider rate limiting this message.

File: code/WorkInProgress/artemis/engine_components.dm [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	desc = "engine casing"
+	icon = 'icons/obj/items/device.dmi'
+	icon_state = "powersink0"
+
+	var/integrity = 0
+	///full integrity for tgui purposes
+	var/full_integrity
+	var/degredation_rate = 0
+
+/obj/item/artemis_engine_component/casing/setup_material()
```
Comment (Azrun) [inline code/WorkInProgress/artemis/engine_components.dm:38]: `onMaterialChanged()`?

File: code/WorkInProgress/artemis/engine_components.dm [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+/obj/item/artemis_engine_component/coil
+	name = "engine coil"
+	desc = "engine coil"
+	icon = 'icons/obj/items/device.dmi'
+	icon_state = "scanner"
+	var/field_strength = 0
+	// var/malfunction_prob
+
+/obj/item/artemis_engine_component/coil/setup_material()
```
Comment (Azrun) [inline code/WorkInProgress/artemis/engine_components.dm:56]: `onMaterialChanged()`?

File: code/WorkInProgress/artemis/engine_components.dm [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+/obj/item/artemis_engine_component/plasma_exciter
+		name = "engine ionizer"
+		desc = "engine ionizer"
+		icon = 'icons/obj/items/device.dmi'
+		icon_state = "signaller"
+
+		var/conversion_rate = 0
+
+/obj/item/artemis_engine_component/plasma_exciter/setup_material()
```
Comment (Azrun) [inline code/WorkInProgress/artemis/engine_components.dm:24]: `onMaterialChanged()`?

File: code/WorkInProgress/artemis/engine_components.dm [hunk 8]
```diff
@@ -0,0 +1,62 @@
+#if ENABLE_ARTEMIS
+ABSTRACT_TYPE(/obj/item/artemis_engine_component)
+/obj/item/artemis_engine_component
+		name = "default engine component"
+		desc = "OUUUUUUGHHGHHHHGHGHHGH(you shouldn't be seeing this please report this)"
+		icon = 'icons/obj/items/device.dmi'
+		icon_state = "signaller"
+
+/obj/item/artemis_engine_component/proc/setup_material()
```
Comment (Azrun) [inline code/WorkInProgress/artemis/engine_components.dm:9]: Should use onMaterialChanged()

File: code/WorkInProgress/artemis/engine_components.dm [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+/obj/item/artemis_engine_component/New(material_name = "steel")
+		..()
+		src.setMaterial(getMaterial(material_name))
+		setup_material()
```
Comment (Azrun) [inline code/WorkInProgress/artemis/engine_components.dm:14]:
```suggestion
```

File: maps/config/map.dm [hunk 10]
```diff
@@ -99,7 +99,7 @@
 #else
 #include "standard.dm"
 #endif
-
+//todo CHANGE THIS BACK BEFORE COMMIT OK?
```
Comment (Azrun) [inline maps/config/map.dm:102]:
```suggestion
```

File: code/WorkInProgress/artemis/artemis_engine.dm [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+/obj/machinery/atmospherics/unary/ion_drive/plasma_tank
+		name = "plasma tank"
+		desc = "the engines plasma tank"
+		icon = 'icons/misc/artemis/artemis_engine.dmi'
+		icon_state = "engine_tank"
+		density = 1
+
+		var/max_pressure =150 * ONE_ATMOSPHERE
+		var/on = FALSE
+		var/volume=1000
```
[the remark this answers is not in the document]
Comment (soapcyst) [reply code/WorkInProgress/artemis/artemis_engine.dm:13]: changed to define. volume is needed to for calculations using the ideal gas equation

File: code/WorkInProgress/artemis/artemis_engine.dm [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+		if(!src.on)
+				return FALSE
+		src.transfer_fuel(src.artemis_tank.stored_fuel,src.artemis_drive.fuel_buffer)
+
+
+/obj/machinery/ion_drive/interface/proc/process_move()
+		if(!src.on)
+				return FALSE
+		if(!src.artemis_drive.usefuel(src.ship.full_throttle))
+				src.on=FALSE;
```
[the remark this answers is not in the document]
Comment (soapcyst) [reply code/WorkInProgress/artemis/artemis_engine.dm:167]: done

File: code/WorkInProgress/artemis/artemis_engine.dm [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+								. = FALSE
+
+				if("toggle-tank")
+						src.artemis_tank.on = !src.artemis_tank.on
+						. = TRUE
+
+				if("adjust-flowrate")
+						var/new_target_pressure = params["target_fuel"]
+						if(isnum(new_target_pressure))
+								src.target_pressure = clamp(new_target_pressure, 0, 10*ONE_ATMOSPHERE)
```
[the remark this answers is not in the document]
Comment (soapcyst) [reply code/WorkInProgress/artemis/artemis_engine.dm:220]: no longer necessary as fuel buffer was removed in preference for using fuel tank directly

**Pull request closed without merging**